src/modm/driver/touch/touch2046.hpp

```diff
@@ -0,0 +1,134 @@
+// coding: utf-8
+/*
+ * Copyright (c) 2021, Raphael Lehmann
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef MODM_TOUCH2046_HPP
+#define MODM_TOUCH2046_HPP
+
+#include <modm/architecture/interface/spi_device.hpp>
+#include <modm/architecture/interface/gpio.hpp>
+#include <modm/processing/resumable.hpp>
+
+namespace modm
+{
+
+/// \ingroup modm_driver_touch2046
+struct touch2046 {
+    enum class Orientation {
+        Normal,
+        //...
+    };
+
+	/**
+	 * \brief	Calibration values are used to calculate touch point
+	 *  from raw values.
+	 *
+	 * \ref FactorX and \ref FactorY scaled by 1000000 to avoid float
+	 *  arithmetic. E.g. to get a factor of 0.75 \ref FactorX has to be
+	 *  set to 750'000.
+	 *
+	 * isTouched() = bool(Z > TresholdZ)
+	 *
+	 * X = (rawX * FactorX / 1000000) + OffsetX
+	 * limited to [0, MaxX]
+	 * Y = (rawY * FactorY / 1000000) + OffsetY
+	 * limited to [0, MaxY]
+	 *
+	 * Orientation (rotation, mirror) are applied after the
+	 * above operations.
+	 */
+	struct Calibration
+	{
+		int16_t OffsetX = 0;
+		int16_t OffsetY = 0;
+		int32_t FactorX = 1'000'000;
+		int32_t FactorY = 1'000'000;
+		uint16_t MaxX = 240;
+		uint16_t MaxY = 320;
+		uint16_t TresholdZ = 1500;
+		Orientation orientation = Orientation::Normal;
+	};
+};
+
+/**
+ * \ingroup	modm_driver_touch2046
+ * \author	Raphael Lehmann
+ *
+ * Datasheet TSC2046: https://www.ti.com/lit/ds/symlink/tsc2046.pdf
+ */
+template < class SpiMaster, class Cs>
+class Touch2046 : public touch2046, public modm::SpiDevice< SpiMaster >, protected modm::NestedResumable<3>
+{
+public:
+	/**
+	 * \brief	Set calibration data
+	 *
+	 * See \ref Calibration.
+	 */
+	void
+	setCalibration(Calibration calibration) {
+		cal = calibration;
+	}
+
+	/**
+	 * \brief	Get raw X, Y and Z values
+	 *
+	 * \return	Position and intensity of touch point. Full int16_t range.
+	 */
+	modm::ResumableResult<std::tuple<uint16_t,uint16_t,uint16_t>>
```

Comment:
There is also `modm::glcd::Point` (which is a typedef of `Vector<int16_t, 2>`), but since modm has no real graphics subsystem (yet?) the user write something himself or integrate third party library like [lvgl](https://lvgl.io/).
For that use-cases I consider `std::tuple` the most useful option, especially thanks to the existence of `std::tie()`.

My impression is also that the modm containers are used less and less, especially since we have [avr-libstdcpp](https://github.com/modm-io/avr-libstdcpp).
I think we could deprecate the modm containers at some point in the future and use STL (where possible) or [ETL](https://github.com/ETLCPP/etl/) instead.

Reply:
+1 for deprecation in favor of ETL.

Reply:
But ETL will not happen in this PR and probably not in the next week.

So is `std::tuple` ok here or should I change it?

Reply:
I'd say leave it for now. Here it is quite clear that the values are coordinates and ordered x,y,z. I've seen real bugs caused by questionable APIs returning tuples with two `int`s and three `bool`s challenging users to get the order right. That's why I got a bit triggered by the extensive use of `std::tuple` when I saw it first but in this case there is no actual risk.